include/zephyr/sys/predef_heap.h

```diff
@@ -0,0 +1,166 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/*
+ * Copyright (c) 2023 Intel Corporation
+ *
+ * Author: Adrian Warecki <adrian.warecki@intel.com>
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_PREDEF_HEAP_H_
+#define ZEPHYR_INCLUDE_SYS_PREDEF_HEAP_H_
+
+#include <zephyr/types.h>
+#include <zephyr/spinlock.h>
+#include <zephyr/sys/util.h>
+
+/* Simple heap implementation with predefined buffers size.
+ *
+ * The blob of memory passed to the heap will be divided into a predefined number of buffers
+ * of specified sizes. Buffers of various sizes are supported. Configuration of a memory division
+ * for a buffers can be changed in runtime. The allocator tries to allocate the smallest buffer that
+ * satisfies the request. Adjacent buffers are not merged, which eliminates the problem of memory
+ * fragmentation. Allocate and free by user pointer and not an opaque block handle. Current
+ * configuration and bookkeeping is stored in a separate memory.
+ */
+
+/* Configuration of a single bundle of buffers. Defines its count and size. */
+struct predef_heap_config {
+	size_t size;
+	int count;
+};
+
+/* predef_heap object structure */
+struct predef_heap {
+	int bundles_count;
+	void *start;
+	size_t size;
+	/* Pointer and size to memory used for configuration store and bookkeeping. */
+	void *config;
```

Comment:
Why `config` struct member is of type void*?
Is it supposed to be used anywhere else by some external API?
It looks like it is used only in your newly added code and you cast it back to `struct predef_heap_bundle*` everywhere.

Reply:
This is a kind of poor-man protection against modification of the configuration from the outside.  
Initially, the `struct predef_heap_bundle` was not supposed to be in the header file, but I found its size to be useful in `PREDEFINED_HEAP_DEFINE` macro. As a user, you enter here the address of the memory blob and its size without interfering with what it is internally.